libgeopmd/src/PowercapSysfsDriver.cpp

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2015 - 2025 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "PowercapSysfsDriver.hpp"
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <cmath>
+#include <cstring>
+#include <sstream>
+
+#include "geopm/Helper.hpp"
+
+#include "SysfsIOGroup.hpp"
+#include "geopm/PlatformTopo.hpp"
+
+static const std::string POWERCAP_DIRECTORY = "/sys/class/powercap";
+
+static std::map<std::string, std::string> load_powercap_resource_by_name(const std::string &powercap_directory)
+{
+    std::map<std::string, std::string> result;
+
+    for (const auto &policy_file : geopm::list_directory_files(powercap_directory)) {
+        if (policy_file.find("intel-rapl:") != 0) {
+            continue;
+        }
+
+        std::ostringstream oss;
+        oss << powercap_directory << "/" << policy_file << "/name";
+        auto name_path = oss.str();
+
+        int name_fd = open(name_path.c_str(), O_RDONLY);
+        if (name_fd == -1) {
+            throw geopm::Exception("PowercapSysfsDriver failed to open " + name_path,
+                                   errno, __FILE__, __LINE__);
+        }
+        char name_buf[geopm::SysfsDriver::M_IO_BUFFER_SIZE] = {};
+        int read_bytes = pread(name_fd, name_buf, sizeof(name_buf) - 1, 0);
+        close(name_fd);
+        if (read_bytes < 0) {
+            throw geopm::Exception("PowercapSysfsDriver failed to read " + name_path,
+                                   errno, __FILE__, __LINE__);
+        }
+        if (read_bytes == sizeof(name_buf) - 1) {
+            throw geopm::Exception("PowercapSysfsDriver truncated read from " + name_path,
+                                   errno, __FILE__, __LINE__);
+        }
```

Comment:
I'm seeing there is opportunity to factor out common code with `CpufreqSysfsDriver`, since I'm seeing the good majority of the code in this function is practically the same. At the very least, this section of code could be factored out into a helper function somewhere that both drivers could implement.

Reply:
Although there are similarities between load_powercap_resource_by_name() and load_cpufreq_cpus_by_resource(), there are some key differences.  The code that is common, is that we are opening all of the files in a directory, reading the contents and doing something depending on the contents.  Much of the work to do that is already factored in geopm::list_directory_files().  

The key differences are the use of the "name" file in the case of powercap, and "affected_cpus" in the cpufreq case to understand the topology.  These file contents are quite different.  I don't see a really easy way to share the code without using std::functions() to be applied to the contents of the files, but that seems overly complex and difficult to read.